kuma/wiki/templatetags/ssr.py

```diff
@@ -0,0 +1,128 @@
+import json
+import subprocess
+import sys
+import time
+
+from django_jinja import library
+
+# A global that holds the reference to a Node child process that does
+# server-side rendering. This will be initialized when first
+# needed. But because of startup overhead we don't want to create this
+# process each time, so we need to keep a referece here so that
+# multiple calls to server_side_render() can use it.
+renderer = None
+
+# The output of the {{render_react_app(data)}} is going to look
+# like this, with optional HTML and state replacing the `{}` markers.
+OUTPUT_TEMPLATE = (
+    u'<div id="react-container">{}</div>\n'
+    u'<script>window._document_data = {};</script>\n'
+)
+
+
+@library.global_function
+def render_react_app(data, ssr=True):
+    """
+    Render a script tag to define the data and any other HTML tags needed
+    to enable the display of a React-based UI. By default, this does
+    server side rendering, falling back to client-side rendering if
+    the SSR attempt fails. Pass False as the second argument to do
+    client-side rendering unconditionally.
+
+    Note that we are not defining a generic Jinja template tag here.
+    The code in this file is specific to Kuma's React-based UI.
+    """
+    if ssr:
+        return server_side_render(data)
+    else:
+        return client_side_render(data)
+
+
+def client_side_render(data):
+    """
+    Output an empty <div> and a script with complete state so that
+    the UI can be rendered on the client-side.
+    """
+    # Convert to JSON and be sure the string "</script>" does not appear
+    serializedState = json.dumps(data).replace('</', '<\\/')
+    return OUTPUT_TEMPLATE.format('', serializedState)
+
+
+def _get_renderer():
+    global renderer
```

Comment:
All of this is making me nervous. 

First of all, it's a pattern I'm just not very familiar with. Secondly, this pattern of using `stdin` and `stdout` is blocking and I don't know what that means if individual Gunicorn (Meinheld) worker use threads. 
Thirdly, perhaps I'm barking up the wrong tree and that all this subprocess stuff will be replaced by a Node HTTP server (and `requests.get()` on the Python side)  before any of this goes into Prod.

Reply:
I can't say I'm not nervous about this too.

1) Are you nervous about the global variable? Or the fact that I'm using a long-lived Node process? Would you be less nervous if I fired up the node process to render a single file and then it exited each time? There would be a lot more overhead, but maybe less risk?

2) Isn't blocking the norm in python code? Aren't we already blocking all over the place in Kuma? What am I missing here?

3) I'm not qualified to change our production infrastructure to introduce a Node-based server into the mix. If we decide to do it that way, I'd be happy to write the server, though!